director/projects/api/views/snapshots.py

```diff
@@ -0,0 +1,141 @@
+from django.conf import settings
+from django.db import IntegrityError
+from django.shortcuts import get_object_or_404
+from django.utils import timezone
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import mixins, permissions, status, viewsets
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from general.api.exceptions import ConflictError
+from lib.data_cleaning import logged_in_or_none
+from projects.models import Snapshot
+from projects.project_models import ProjectEvent, ProjectEventType, ProjectEventLevel
+from projects.snapshots import ProjectSnapshotter, SnapshotInProgressError
+from projects.views.mixins import ProjectPermissionsMixin, ProjectPermissionType
+from projects.api.serializers import SnapshotSerializer, SnapshotCreateRequestSerializer
+
+
+class SnapshotsViewSet(
+    mixins.ListModelMixin,
+    mixins.CreateModelMixin,
+    mixins.RetrieveModelMixin,
+    viewsets.GenericViewSet,
+    ProjectPermissionsMixin,
+):
+
+    serializer_class = SnapshotSerializer
+    lookup_field = "number"
+
+    def get_permissions(self):
```

Comment:
Is this method required given we check the permissions more specifically inside the `list` and `create` methods anyway?

Reply:
I agree it seems a little redundant. But it does provide an opportunity to "deny early" (i.e. before getting to those other methods) and returns a 401 code instead of 403, thereby informing the client of the need to authenticate. Also, when it comes to security I think redundancy like this is probably a good thing.